### PR TITLE
Mobile wallet feature parity

### PR DESCRIPTION
**Pull request by soerenbf**
## Changes

- Adds support for deriving encryption keys + combining and decrypting encrypted amounts
- Add support for ID2.5 proofs and verifiable credentials + presentations

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: build.rs [hunk 1]
```diff
@@ -1,3 +1,15 @@
...
+    let out_dir = std::env::var_os("OUT_DIR").unwrap();
+    let dest_path = std::path::Path::new(&out_dir).join("table_bytes.bin");
+
+    let global = id::types::GlobalContext::<concordium_base::id::constants::ArCurve>::generate(
+        String::from("genesis_string"),
+    );
+    let m = 1 << 16;
+    let table = elgamal::BabyStepGiantStep::new(global.encryption_in_exponent_generator(), m);
+    std::fs::write(dest_path, common::to_bytes(&table)).expect("Could not write table to file.");
+
```
Comment (DOBEN) [inline build.rs:13]: Just curious: is `genesis_string` the seed for testnet and mainnet for the cryptographic parameters?

File: src/id_proofs.rs [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    /// The attribute that the verifier wants the user to prove is in a range.
+    #[serde(rename = "attributeTag")]
+    pub attribute_tag: Tag,
+    /// The lower bound on the range.
+    #[serde(rename = "lower")]
+    pub lower: Value,
+    #[serde(rename = "upper")]
+    /// The upper bound of the range.
+    pub upper: Value,
+}
```
Comment (DOBEN) [inline src/id_proofs.rs]: Specifying `camelCase` can save some of the write-up. This applies to other large structs as well.

```suggestion
#[derive(Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")] 
pub struct AttributeInRangeStatement<Tag: Clone, Value: Clone> {
    /// The attribute that the verifier wants the user to prove is in a range.
    pub attribute_tag: Tag,
    /// The lower bound on the range.
    pub lower: Value,
    /// The upper bound of the range.
    pub upper: Value,
}
```

File: src/id_proofs.rs [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+/// commitment. Since the verifier does not know the attribute value before
+/// seing the proof, the value is not present here.
+#[derive(Serialize, Deserialize, Clone)]
+pub struct RevealAttributeStatement<Tag: Clone> {
+    /// The attribute that the verifier wants the user to reveal.
+    #[serde(rename = "attributeTag")]
+    pub attribute_tag: Tag,
+}
+
+/// Serves as a uniFFI compatible bridge to [`concordium_base::id::id_proof_types::RevealAttributeStatement<ArCurve, AttributeTag>`]
```
Comment (DOBEN) [inline src/id_proofs.rs]:
```suggestion
/// Serves as a uniFFI compatible bridge to [`concordium_base::id::id_proof_types::RevealAttributeStatement<AttributeTag>`]
```

File: src/id_proofs.rs [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    pub upper: Value,
+}
+
+/// Serves as a uniFFI compatible bridge to [`concordium_base::id::id_proof_types::AttributeInRangeStatement<ArCurve, AttributeTag, AttributeKind>`]
+pub type AttributeInRangeStatementV1 = AttributeInRangeStatement<AttributeTag, String>;
+
+/// For the case where the verifier wants the user to prove that an attribute is
+/// in a set of attributes.
+///
+/// Serves as a uniFFI compatible bridge to [`concordium_base::id::id_proof_types::AttributeInSetStatement<ArCurve, AttributeTag, AttributeKind>`]
```
Comment (DOBEN) [inline src/id_proofs.rs]:
```suggestion

```

File: src/id_proofs.rs [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    pub set: Vec<Value>,
+}
+
+/// Serves as a uniFFI compatible bridge to [`concordium_base::id::id_proof_types::AttributeInSetStatement<ArCurve, AttributeTag, AttributeKind>`]
+pub type AttributeInSetStatementV1 = AttributeInSetStatement<AttributeTag, String>;
+
+/// For the case where the verifier wants the user to prove that an attribute is
+/// not in a set of attributes.
+///
+/// Serves as a uniFFI compatible bridge to [`concordium_base::id::id_proof_types::AttributeNotInSetStatement<ArCurve, AttributeTag, AttributeKind>`]
```
Comment (DOBEN) [inline src/id_proofs.rs]:
```suggestion

```

File: src/id_proofs.rs [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+    fn try_from(value: StatementV1) -> Result<Self, Self::Error> {
+        serde_json::to_string(&value).and_then(|s| serde_json::from_str(&s))
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(tag = "type")]
+pub enum AtomicProof<Value> {
+    /// The atomic statement stating that an attribute should be revealed.
```
Comment (DOBEN) [inline src/id_proofs.rs]: The attribute is already revealed at this point.

File: src/id_proofs.rs [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+}
+
+/// Prove a [`StatementV1`] given the provided context, producing a [`VersionedProofV1`]
+#[allow(clippy::too_many_arguments)]
+pub fn prove_statement_v1(
+    seed: Bytes,
+    net: Network,
+    global_context: GlobalContext,
+    ip_index: u32,
+    identity_index: u32,
```
Comment (DOBEN) [inline src/id_proofs.rs:176]: Suggestion: I feel we call it usually `key_index`, but I find `identity_index` might be even better to distinguish the value from other `keys`, so kind of like that as well.

File: src/id_proofs.rs [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+        )
+        .map_err(|e| e.to_call_failed(fn_name.to_string()))?;
+    let statement = StatementWithContext {
+        statement,
+        credential: cred_id,
+    };
+
+    let proof = statement
+        .prove(
+            ProofVersion::Version1,
```
Comment (DOBEN) [inline src/id_proofs.rs]: Why don't we use `Version2` proofs here (the Fiat-Shamir heuristic was re-written and this version of proofs should be used going forward)?

File: src/id_proofs.rs [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+            &global_context,
+            challenge.as_ref(),
+            &id_object.alist,
+            &credential_context,
+        )
+        .context("Could not produce proof.")
+        .map_err(|e| e.to_call_failed(fn_name.to_string()))?;
+    VersionedProofV1::try_from(concordium_base::common::Versioned::new(VERSION_0, proof))
```
Comment (DOBEN) [inline src/id_proofs.rs]: A bit confusing that we started to call the struct `VersionedProofV1` and then have a corresponding constant `VERSION_0` in base.

File: src/lib.udl [hunk 10]
```diff
@@ -367,56 +441,349 @@ dictionary TypeSchema {
...
+    /// The secret key serialized as bytes.
+    Bytes secret;
+    /// The public key serialized as bytes.
+    Bytes public;
+};
+
+/// For the case where the verifier wants the user to show the value of an
+/// attribute and prove that it is indeed the value inside the on-chain
+/// commitment. Since the verifier does not know the attribute value before
+/// seing the proof, the value is not present here.
```
Comment (DOBEN) [inline src/lib.udl]:
```suggestion
/// seeing the proof, the value is not present here.
```

File: src/lib.udl [hunk 11]
```diff
@@ -367,56 +441,349 @@ dictionary TypeSchema {
...
+
+/// A statement is a list of atomic statements.
+dictionary StatementV1 {
+    sequence<AtomicStatementV1> statements;
+};
+
+/// The different types of proofs, corresponding to the statements above.
+[Enum]
+interface AtomicProofV1 {
+    /// The atomic statement stating that an attribute should be revealed.
```
Comment (DOBEN) [inline src/lib.udl]: The attribute is already revealed

Comment (DOBEN) [inline src/lib.udl]:
```suggestion
    /// The atomic proof stating that an attribute should be revealed.
```

File: src/lib.udl [hunk 12]
```diff
@@ -367,56 +441,349 @@ dictionary TypeSchema {
...
+    sequence<AtomicProofV1> proofs;
+};
+
+/// A versioned variant of `ProofV1`
+dictionary VersionedProofV1 {
+    /// The version associated with the value
+    u32 version;
+    /// The proof
+    ProofV1 value;
+};
```
Comment (DOBEN) [inline src/lib.udl]: It is called proofV1 and then it tracks another version in it. Can we guarantee that they match?

File: src/lib.udl [hunk 13]
```diff
@@ -367,56 +441,349 @@ dictionary TypeSchema {
...
+/// A value of an attribute. This is the low-level representation. The
+/// different variants are present to enable different representations in JSON,
+/// and different embeddings as field elements when constructing and verifying
+/// proofs.
+[Enum]
+interface Web3IdAttribute {
+    String(string value);
+    Numeric(u64 value);
+    Timestamp(timestamp value);
+};
```
Comment (DOBEN) [inline src/lib.udl:547]:
```suggestion
};

```

File: src/lib.udl [hunk 14]
```diff
@@ -367,56 +441,349 @@ dictionary TypeSchema {
...
+/// comes separately.
+dictionary VerifiablePresentationRequest {
+    Bytes challenge;
+    sequence<VerifiableCredentialStatement> statements;
+};
+
+/// The different types of proofs, corresponding to the statements above.
+[Enum]
+interface AtomicProofV2 {
+    /// The atomic statement stating that an attribute should be revealed.
```
Comment (DOBEN) [inline src/lib.udl]:
```suggestion
    /// The atomic proof containing the revealed attributes.
```

File: src/lib.udl [hunk 15]
```diff
@@ -499,4 +866,33 @@ namespace crypto {
     /// Get the `TypeSchema` for events emitted by a contract
     [Throws=SchemaError]
     TypeSchema get_event_schema(ModuleSchema schema, string contract_name);
+
+    /// Attempt to create the encryption keys associated with an account at the given credential index
+    [Throws=ConcordiumWalletCryptoError]
+    EncryptionKeys get_encryption_keys(GlobalContext global_context, Bytes prf_key, u8 credential_index);
+
+    /// Attempt to create the encryption keys associated with an account at the given credential index
```
Comment (DOBEN) [inline src/lib.udl]: ?

File: src/id_proofs.rs (hunk 8 above)
Comment (DOBEN) [reply src/id_proofs.rs]: From my understanding, the proof version has nothing to do with the statement version here. Meaning we have `statementVersion1` but that does not mean we need to use `ProofVersion1` here.

File: src/lib.udl [hunk 16]
```diff
@@ -499,4 +866,33 @@ namespace crypto {
...
+    [Throws=ConcordiumWalletCryptoError]
+    MicroCCDAmount decrypt_amount(Bytes encrypted_amount, Bytes encryption_secret_key);
+
+    /// Attempt to combine two encrypted amounts
+    [Throws=ConcordiumWalletCryptoError]
+    Bytes combine_encrypted_amounts(Bytes left, Bytes right);
+
+    /// Prove the supplied statement in the context of a given identity and credential
+    [Throws=ConcordiumWalletCryptoError]
+    VersionedProofV1 prove_statement_v1(Bytes seed, Network net, GlobalContext global_context, u32 ip_index, u32 identity_index, u8 credential_index, IdentityObject identity_object, StatementV1 statement, Bytes challenge);
```
Comment (DOBEN) [inline src/lib.udl]: The newest proof version is V2, so while we prove statement v1, it might be best to create a proof version 2 here. E.g.
```
VersionedProofV2 prove_statement_v1
```

File: src/web3id.rs [hunk 17]
```diff
@@ -0,0 +1,797 @@
+use std::{
+    collections::{BTreeSet, HashMap},
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+use chrono::{DateTime, Utc};
```
Comment (DOBEN) [inline src/web3id.rs]: Do we need `std::time::SystemTime` and `UNIX_EPOCH` from `std` or could we rely on chrono exclusively for time calculations? Since you convert `SystemTime` into a chrono type later and the chrono type is human readable, I suggest to just use the chrono type. 

I am worried that there are slight differences in the definition of times between libraries and that can cause problems down the line. In the base libraries we rely on `chrono`.

File: src/web3id.rs [hunk 18]
```diff
@@ -0,0 +1,797 @@
...
+        match value {
+            Web3IdAttribute::String { value } => {
+                web3id::Web3IdAttribute::String(AttributeKind(value.to_string()))
+            }
+            Web3IdAttribute::Numeric { value } => web3id::Web3IdAttribute::Numeric(value),
+            Web3IdAttribute::Timestamp { value } => {
+                let v = DateTime::<Utc>::from(value);
+                web3id::Web3IdAttribute::Timestamp(Timestamp {
+                    millis: v.timestamp_millis() as u64,
+                })
```
Comment (DOBEN) [inline src/web3id.rs]: I think we should stick to the `chrono::DateTime` instead of converting between `std::time::SystemTime`.

I also want to highlight that there was a breaking change in `chrono` library (versions of chrono (<0.32)) regarding their time constants that caused us to introduce an offset to align the time to the old serialization standard.
https://github.com/Concordium/concordium-base/blob/13048aa52e424f854790945265054412e1d24b3d/rust-src/concordium_base/src/web3id/mod.rs#L1677

The above `try_from` chrono::DateTime implementation should be used instead.

File: src/web3id.rs [hunk 19]
```diff
@@ -0,0 +1,797 @@
...
+        match value {
+            web3id::Web3IdAttribute::String(value) => Web3IdAttribute::String { value: value.0 },
+            web3id::Web3IdAttribute::Numeric(value) => Web3IdAttribute::Numeric { value },
+            web3id::Web3IdAttribute::Timestamp(value) => Web3IdAttribute::Timestamp {
+                value: UNIX_EPOCH + std::time::Duration::from_millis(value.millis),
+            },
+        }
+    }
+}
+/// Serves as a uniFFI compatible bridge to [`concordium_base::id::id_proof_types::RevealAttributeStatement<ArCurve, String>`]
```
Comment (DOBEN) [inline src/web3id.rs]:
```suggestion
/// Serves as a uniFFI compatible bridge to [`concordium_base::id::id_proof_types::RevealAttributeStatement<String>`]
```

File: src/id_proofs.rs [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+    /// The atomic proof stating that an attribute is in a range.
+    AttributeInRange { proof: Bytes },
+    /// The atomic proof stating that an attribute is in a set.
+    AttributeInSet { proof: Bytes },
+    /// The atomic proof stating that an attribute is not in a set.
+    AttributeNotInSet { proof: Bytes },
+}
+
+/// Serves as a uniFFI compatible bridge to [`concordium_base::id::id_proof_types::AtomicProof<ArCurve, AttributeKind>`]
+pub type AtomicProofV1 = AtomicProof<String>;
```
Comment (DOBEN) [inline src/id_proofs.rs]: Can we add some test cases to the `web3id` file that proof and verify, please. In particular, I would be interested if the `String` works for the attribute `sex` since it is a number ISO/IEC 5218 and range proofs with a `date-time` attribute.

I wonder why `String` was used for identity proofs in this PR compared to the original implementation.

File: src/web3id.rs [hunk 21]
```diff
@@ -0,0 +1,797 @@
...
+        /// [`concordium_base::base::CredentialRegistrationID`]
+        cred_id: Bytes,
+        statement: Vec<AtomicStatementV1>,
+    },
+    /// Statement about a credential issued by a Web3 identity provider, a smart
+    /// contract.
+    Web3Id {
+        /// The credential type. This is chosen by the provider to provide
+        /// some information about what the credential is about.
+        cred_type: Vec<String>,
```
Comment (DOBEN) [inline src/web3id.rs:112]: Is there a benefit for using a `vector` over the original `BTreeSet`?

File: src/web3id.rs [hunk 22]
```diff
@@ -0,0 +1,797 @@
...
+{
+    type Error = serde_json::Error;
+
+    fn try_from(value: VerifiableCredentialCommitmentInputs) -> Result<Self, Self::Error> {
+        serde_json::to_value(value).and_then(serde_json::from_value)
+    }
+}
+
+/// Serves as a uniFFI compatible bridge to [`id::id_proof_types::AtomicProof<ArCurve, Web3IdAttribute>`]
+pub type AtomicProofV2 = AtomicProof<Web3IdAttribute>;
```
Comment (DOBEN) [inline src/web3id.rs]: There exist `proofV1` and `proofV2` (which have different implementations of the fiat-shamir heuristic).

The `AtomicProofV1` relates to proofs generated from the identity (not to the above different proof versions)  = > Naming wise I suggest: `AtomicProofIdentity`
`AtomicProofV2` relates here to proofs generated from web3id credential. = > Naming wise I suggest: `AtomicProofWeb3Credential`

`StatementV1` and `StatementV2` naming could be adjusted but not feeling strongly about it.

File: src/web3id.rs [hunk 23]
```diff
@@ -0,0 +1,797 @@
...
+
+/// A pair of a statement and a proof.
+#[derive(serde::Deserialize)]
+#[serde(from = "(AtomicStatement<Tag, Value>, AtomicProof<Value>)")]
+pub struct CredentialStatementWithProof<Tag: Clone, Value: Clone> {
+    pub statement: AtomicStatement<Tag, Value>,
+    pub proof: AtomicProof<Value>,
+}
+
+/// Serves as a uniFFI compatible bridge to [`web3id::StatementWithProof<ArCurve, String, String>`]
```
Comment (DOBEN) [inline src/web3id.rs]:
```suggestion
/// Serves as a uniFFI compatible bridge to [`web3id::StatementWithProof<ArCurve, AttributeTag, String>`]
```

File: src/web3id.rs [hunk 24]
```diff
@@ -0,0 +1,797 @@
...
+    let fn_desc = format!("parse_did_method(value={})", value);
+    let method = web3id::did::Method::try_from(value)
+        .context("Failed to parse DID")
+        .map_err(|e| e.to_call_failed(fn_desc.to_string()))?;
+    Ok(method.into())
+}
+
+/// Format the given `Method` as a DID string
+pub fn did_method_as_string(did: DID) -> Result<String, ConcordiumWalletCryptoError> {
+    let fn_desc = format!("parse_did_method(value={:?})", did);
```
Comment (DOBEN) [inline src/web3id.rs]:
```suggestion
    let fn_desc = format!("did_method_as_string(value={:?})", did);
```

File: src/web3id.rs [hunk 25]
```diff
@@ -0,0 +1,797 @@
...
+        let commitment_inputs = VerifiableCredentialCommitmentInputs::Web3Issuer {
+            signature: Bytes(hex::decode("40ced1f01109c7a307fffabdbea7eb37ac015226939eddc05562b7e8a29d4a2cf32ab33b2f76dd879ce69fab7ff3752a73800c9ce41da6d38b189dccffa45906").unwrap()),
+            signer: hex::decode("5c860c2333dceca06d6de233171c7daeb5f6f0af3686d9130e02dbe941f4db3a").unwrap(),
+            values: values.into_iter().collect(),
+            randomness: randomness.into_iter().collect(),
+        };
+        web3id::OwnedCommitmentInputs::try_from(commitment_inputs.clone())
+            .expect("Can convert commitment inputs to base type");
+
+        create_verifiable_presentation(request, get_global(), vec![commitment_inputs])
```
Comment (DOBEN) [inline src/web3id.rs:836]: A nice to have would be a few test cases that create a verifiable presentation and verify it, to ensure it works.

File: src/web3id.rs [hunk 26]
```diff
@@ -0,0 +1,797 @@
...
+                ],
+            }],
+        };
+
+        web3id::Request::try_from(request.clone()).expect("Can convert request to base type");
+
+        let values: Vec<(String, Web3IdAttribute)> = vec![
+            (
+                "degreeName".to_string(),
+                Web3IdAttribute::String {
```
Comment (DOBEN) [inline src/web3id.rs:781]: A nice to have would be a mix of web3id credentials of different attributes `string`, `date-time` and `number`, and the same for the identity proof e.g. sex (`number`), firstName (`string`), and dob (`data-time`).

File: src/web3id.rs [hunk 27]
```diff
@@ -0,0 +1,797 @@
...
+                                Web3IdAttribute::String {
+                                    value: "BachelorDegree".to_string(),
+                                },
+                                Web3IdAttribute::String {
+                                    value: "MasterDegree".to_string(),
+                                },
+                            ],
+                        },
+                    },
+                    AtomicStatementV2::RevealAttribute {
```
Comment (DOBEN) [inline src/web3id.rs]: Nice to have would be a test case for range proof here as well for the web3 proofs.

File: src/web3id.rs [hunk 28]
```diff
@@ -0,0 +1,797 @@
...
+        let request = VerifiablePresentationRequest {
+            challenge: Bytes(
+                hex::decode("94d3e85bbc8ff0091e562ad8ef6c30d57f29b19f17c98ce155df2a30100dAAAA").unwrap(),
+            ),
+            statements: vec![
+                VerifiableCredentialStatement::Account {
+                    network: Network::Testnet,
+                    cred_id: Bytes(hex::decode("94d3e85bbc8ff0091e562ad8ef6c30d57f29b19f17c98ce155df2a30100df4cac5e161fb81aebe3a04300e63f086d0d8").unwrap()),
+                    statement: vec![
+                        AtomicStatementV1::AttributeInRange { statement: AttributeInRangeStatementV1 {
```
Comment (DOBEN) [inline src/web3id.rs]: Nice ot have would be a test case for the `revealAttribute` and `inSet/notInSet` proofs for the identity proofs.

File: src/web3id.rs [hunk 29]
```diff
@@ -0,0 +1,797 @@
...
+                        }
+                    ]
+                }
+            ],
+        };
+
+        web3id::Request::try_from(request.clone()).expect("Can convert request to base type");
+
+        let values: Vec<(AttributeTag, String)> = vec![
+            (AttributeTag::DateOfBirth, "0".to_string()),
```
Comment (DOBEN) [inline src/web3id.rs:668]: can we test an attribute with `date-time` one with number e.g. (`sex`) and one with string.

File: src/web3id.rs (hunk 18 above)
Comment (DOBEN) [reply src/web3id.rs]: Adding this test to this PR would give us peace in mind that we still correctly adjust the time to the serialization rule in the old time libraries.

https://github.com/Concordium/concordium-base/blob/13048aa52e424f854790945265054412e1d24b3d/rust-src/concordium_base/src/web3id/mod.rs#L2304

Adding a test case that generates and verifies a proof very close to the boundary of a `date-time` attribute in a range proof would be quite nice to have.

File: build.rs (hunk 1 above)
Comment (soerenbf) [reply build.rs:13]: This struck me as weird as well. It's basically copied from the implementation in the "old" mobile wallet library. The internal `genesis_string` of `GlobalContext` is not used to derive any of the cryptographic functionality required for this build step, which (I assume) is why it's just an arbitrary string.

File: src/id_proofs.rs (hunk 7 above)
Comment (soerenbf) [reply src/id_proofs.rs:176]: Inspiration is mainly from JS SDK with regards to the input for this function, and in the corresponding input there it's called `identityIndex` 🤷 

I personally like `identity_index` more as well since it describes the "ip>identity>credential" path on the derivation function on the seed better.

File: src/lib.udl (hunk 15 above)
Comment (soerenbf) [reply src/lib.udl]: good catch 😄

File: src/web3id.rs (hunk 17 above)
Comment (soerenbf) [reply src/web3id.rs:6]: Good suggestion, will adapt 👍

Comment (soerenbf) [reply src/web3id.rs:6]: Will change this as `SystemTime` cannot represent time prior to unix epoch apparently.. Will use ISO8601 format as input instead which should provide consistent results.

Thanks for pointing this out as it was a potential pitfall.. I don't know why we use `contracts_common::Timestamp` inside `Web3IdAttribute::Timestamp`, as the former is described as not supporting values prior to unix epoch, while the outer interprets the inner timestamp as time since the minimum valid date in chrono (`"-26144-01-01T:00:00:00Z"`) which is quite a significant loss of precision 😅

File: src/web3id.rs (hunk 21 above)
Comment (soerenbf) [reply src/web3id.rs:112]: `BTreeSet` cannot be translated by uniFFI: https://mozilla.github.io/uniffi-rs/latest/udl/builtin_types.html. The types declared as input/output types of the FFI layer must boil down to this set of types

File: src/web3id.rs (hunk 22 above)
Comment (soerenbf) [reply src/web3id.rs]: Yes I agree, will change this 🙂

File: src/web3id.rs (hunk 25 above)
Comment (soerenbf) [reply src/web3id.rs:836]: I don't know, to me this seems redundant as these tests are already in place in `concordium_base::web3id`. The main point of the tests in this library is that the input types and output types properly convert into their concordium_base counterparts.

File: src/id_proofs.rs (hunk 20 above)
Comment (soerenbf) [reply src/id_proofs.rs]: The only supported values for identity attributes are `String` values: https://github.com/Concordium/concordium-base/blob/13048aa52e424f854790945265054412e1d24b3d/rust-src/concordium_base/src/id/constants.rs#L37. This is the assumption for the original ID proofs as well which this type also represents

File: src/web3id.rs (hunk 26 above)
Comment (soerenbf) [reply src/web3id.rs:696]: I'll add tests for converting between the input/output types and the corresponding `concordium_base` types, but IMO other tests will be mostly redundant as this will already have tests in the modules in base.

File: src/web3id.rs (hunk 29 above)
Comment (soerenbf) [reply src/web3id.rs:668]: All identity attributes are represented as strings, so we can't do that I'm afraid 😛 All identity attribute values are strings and are all converted to `concordium_base::web3id::Web3IdAttribute::String`. This is also the reason why `Web3IdAttribute` is not used in the `Account` variants of the different `enum`s

File: src/lib.udl (hunk 12 above)
Comment (soerenbf) [reply src/lib.udl]: This is all due to bad naming.. I wanted to align with the names used in the js sdk, but your suggestion with id/web3id postfixes is better so I went with that.

File: src/web3id.rs (hunk 18 above)
Comment (soerenbf) [reply src/web3id.rs]: Changed conversion to use the conversion from/to `chrono::Datetime` implemented for `concordium_base::web3id::Web3IdAttribute`

File: src/web3id.rs (hunk 26 above)
Comment (soerenbf) [reply src/web3id.rs:781]: I added tests that ensure the conversion to/from `web3id::Web3IdAttribute` works for all variants of `Web3IdAttribute`. This should be sufficient as tests for generating presentations for different variants should be covered in `concordium_base`

Comment (soerenbf) [reply src/web3id.rs:781]: I added the remaining unrepresented attributes to the presentation creation tests as well 😄

File: src/id_proofs.rs (hunk 8 above)
Comment (soerenbf) [reply src/id_proofs.rs]: Changed it to use `ProofVersion::Version2`. Thanks for the suggestion 😄

File: build.rs (hunk 1 above)
Comment (DOBEN) [reply build.rs:13]: Would be nice to add a comment once we get a response from science about how the cryptographic parameters where derived.

File: CHANGELOG.md [hunk 30]
```diff
@@ -5,6 +5,27 @@ All notable changes to this project will be documented in this file.
...
+
+- **BREAKING** - `SecToPubTransferData.transfer_amount` has been changed to `MicroCCDAmount` (previously `String`) which is an alias in generated code for `u64`
+- **BREAKING** - Any place where the `String` identifier of a concordium network was used, the new `Network` type is now used.
+- **BREAKING** - Identity attribute tags are now represented by the `AttributeTag` type instead of `u8` and replaces any occurance of this.
+- **BREAKING** - `AccountCredentialResult` is now called `AccountCredentialWithRandomness` to better signal its use, i.e. store the randomness corresponding to a credential.
+
+### Added
+
+- `get_encryption_keys`, `decrypt_amount`, and `combine_encrypted_amounts` functions for handling encrypted amounts.
+- `prove_statement_v1` for producing proofs related to Concordium identities.
```
Comment (DOBEN) [inline CHANGELOG.md]: needs update

File: src/id_proofs.rs [hunk 31]
```diff
@@ -0,0 +1,234 @@
...
+        .prove(
+            ProofVersion::Version2,
+            &global_context,
+            challenge.as_ref(),
+            &id_object.alist,
+            &credential_context,
+        )
+        .context("Could not produce proof.")
+        .map_err(|e| e.to_call_failed(fn_name.to_string()))?;
+    VersionedIdentityProof::try_from(concordium_base::common::Versioned::new(VERSION_0, proof))
```
Comment (DOBEN) [inline src/id_proofs.rs]: I don't fully understand the purpose of this `VERSION_0`. Will it always be 0 for now or is it related to the `proofV2` ?

**Review by DOBEN — APPROVED** [review]
LGTM

File: src/id_proofs.rs (hunk 31 above)
Comment (soerenbf) [reply src/id_proofs.rs]: In the JS SDK, the proof produced for `VERSION_0` is `ProofVersion::Version1`, so the version should (probably) be bumped. The proof version is needed to verify the proof.. I can't find any documentation pointing to a link between a `Versioned<Proof>` and `ProofVersion` 🤷 . Bare in mind this is for the "old" ID proofs and is not used for ID2.5 verifiable credentials so it probably will not be updated.

Have you worked with ID proofs versioned this way?

Comment (DOBEN) [reply src/id_proofs.rs]: `VERSION_0` for `ProofVersion::Version1`
and
`VERSION_1` for `ProofVersion::Version2`
makes sense, so I am fine if we go with that.
